xiaogpt.py

```diff
@@ -74,26 +74,39 @@ async def ask(self, query):
 
 class ChatGPTBot:
     def __init__(self, session):
-        pass
+        self.history = []
 
     async def ask(self, query):
         openai.api_key = OPENAI_API_KEY
-        completion = openai.ChatCompletion.create(
-            model="gpt-3.5-turbo",
-            messages=[
-                {
-                    "role": "user",
-                    "content": f"{query}",
-                }
-            ],
-        )
+        ms = []
+        if len(self.history) >= 5:
+            ms = [
+                {"role": "user", "content": self.history[-5][0]},
+                {"role": "assistant", "content": self.history[-5][1]},
+                {"role": "user", "content": self.history[-4][0]},
+                {"role": "assistant", "content": self.history[-4][1]},
+                {"role": "user", "content": self.history[-3][0]},
+                {"role": "assistant", "content": self.history[-3][1]},
+                {"role": "user", "content": self.history[-2][0]},
+                {"role": "assistant", "content": self.history[-2][1]},
+                {"role": "user", "content": self.history[-1][0]},
+                {"role": "assistant", "content": self.history[-1][1]},
+                {"role": "user", "content": f"{query}"},
+            ]
+        else:
+            for h in self.history:
+                ms.append({"role": "user", "content": h[0]})
+                ms.append({"role": "assistant", "content": h[1]})
+            ms.append({"role": "user", "content": f"{query}"})
+        completion = openai.ChatCompletion.create(model="gpt-3.5-turbo", messages=ms)
```

Comment:
It's recommend to define the model as constant, it will move to the config.json later

Reply:
And how about this style
```
ms = []
if len(self.history) >= 5:
    for i in range(-5, 0):
        ms.append({"role": "user", "content": self.history[i][0]})
        ms.append({"role": "assistant", "content": self.history[i][1]})
else:
    for h in self.history:
        ms.append({"role": "user", "content": h[0]})
        ms.append({"role": "assistant", "content": h[1]})
ms.append({"role": "user", "content": f"{query}"})
completion = openai.ChatCompletion.create(model="gpt-3.5-turbo", messages=ms)

```

Reply:
ok

Reply:
if `self.history` is a list, you can use `self.history[-5:]`

```
ms = []
for h in self.history[-5:]:
    ms.append({"role": "user", "content": h[0]})
    ms.append({"role": "assistant", "content": h[1]})
ms.append({"role": "user", "content": f"{query}"})
completion = openai.ChatCompletion.create(model="gpt-3.5-turbo", messages=ms)
```